cmd/activator/main.go

```
@@ -133,6 +150,84 @@ func main() {
 	}
 
+	// Set up signals so we handle the first shutdown signal gracefully.
+	stopCh := signals.SetupSignalHandler()
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, defaultResyncInterval)
+	servingInformerFactory := servinginformers.NewSharedInformerFactory(servingClient, defaultResyncInterval)
+	endpointInformer := kubeInformerFactory.Core().V1().Endpoints()
+	revisionInformer := servingInformerFactory.Serving().V1alpha1().Revisions()
+
+	go endpointInformer.Informer().Run(stopCh)
+	go revisionInformer.Informer().Run(stopCh)
+
+	params := queue.BreakerParams{QueueDepth: breakerQueueDepth, MaxConcurrency: breakerMaxConcurrency, InitialCapacity: 0}
+
+	// Return the number of endpoints, 0 if no endpoints are found.
+	endpointsGetter := func(revID activator.RevisionID) (int32, error) {
+		endpointKey := cache.ExplicitKey(revID.Namespace + "/" + reconciler.GetServingK8SServiceNameForObj(revID.Name))
+		endpoints, exists, err := endpointInformer.Informer().GetIndexer().Get(endpointKey)
+		if err != nil {
+			return 0, fmt.Errorf("Error getting endpoints for revision %v", revID)
+		}
+		if !exists {
+			// No endpoints exist yet.
+			return 0, nil
+		}
+		if exists && endpoints != nil {
+			endpoint := endpoints.(*corev1.Endpoints).Subsets
+			addresses := activator.EndpointsAddressCount(endpoint)
+			return int32(addresses), nil
+		}
+		return 0, nil
+	}
+
+	// Return the revision from the observer.
+	// Additionally to observer errors an error is returned if either no revision exists yet
+	// or we could not cast the interface to Revision for some reason.
+	revisionGetter := func(revID activator.RevisionID) (revision *v1alpha12.Revision, err error) {
+		revisionKey := cache.ExplicitKey(revID.Namespace + "/" + revID.Name)
+		rev, ok, err := revisionInformer.Informer().GetIndexer().Get(revisionKey)
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			return nil, fmt.Errorf("Error getting revision for %v", revID)
+		}
```

Do we want to return an error for `!ok` here? If so, we could be more explicit and state that the revision does not exist.

---

I'm going to replace it with the lister, the error should come from it.